### DISCUSSION TITLE
Unable to type decorated @property on instance

### DISCUSSION THREAD

**Discussion by argilya** [microsoft/pyright#3785] [Q&A] [answered]
Hello,

I am learning OOP with Python, and today I ran into a problem trying to type a decorator shared by a hierarchy of classes.

Here is the use case:

I have two classes, `Parent` and a subclass `Child`. Both of them have a property called `data`. When you call `.data`, there is a chance the code will throw an exception. Rather than handle the exception inside `.data`, I wrote a decorator that handles the exception. This way I can decorate `.data` in both `Parent` and any of its `Children` and re-use code. However, I am not able to type the code properly, as Pyright is unable to access `.data`.

Here is an example with just the `Parent` class, where everything works as expected. Notice how Pyright correctly identifies `parent.data` as a `float`.
```python
from __future__ import annotations

from typing import Callable
from typing_extensions import reveal_type


def handle_error(fn: Callable[[Parent], float]):
    def wrapper(instance: Parent) -> float:
        try:
            return fn(instance)
        except ZeroDivisionError: # simulate example error with zero division
            return 0

    return wrapper


class Parent:
    @property
    @handle_error
    def data(self):
        return 1 / 0


parent = Parent()

reveal_type(parent.data)  # float
```

However, things break down when I add a `Child` to `Parent`, and type the decorator to accept any "generic" instance of `Parent`. Here is the code:

```python
from __future__ import annotations

from typing import Callable, TypeVar
from typing_extensions import reveal_type


AnyParent = TypeVar('AnyParent', bound='Parent') # this should mean Parent and all subclasses


def handle_error(fn: Callable[[AnyParent], float]):
    def wrapper(instance: AnyParent) -> float:
        try:
            return fn(instance)
        except ZeroDivisionError:
            return 0

    return wrapper


class Parent:
    @property
    @handle_error
    def data(self):
        return 1 / 0


class Child(Parent):
    @property
    @handle_error
    def data(self):
        return 1


parent = Parent()
child = Child()

reveal_type(parent.data)  # Unknown & Cannot access member "data" ...
reveal_type(child.data)  # Unknown & Cannot access member "data" ...
```

Interestingly, when I remove the `@property` decorators from `.data`, and call the methods with `.data()`, Pyright is able to detect the method and its return.

I am not sure where I am making a mistake. I think, I am either misusing generics or `@property`.  Any feedback is appreciated, thank you!

**Comment by erictraut** [comment, marked as answer]
Properties require a bunch of special-case logic in a Python type checker, and there are certain edge cases that this logic doesn't handle well. The reason this isn't working well here is because `@property` assumes that the function it is decorating is an unbound instance method — i.e. an instance method with a `self` parameter. Type checkers like Pyright internally track whether a function is an instance method. However, when you pass such a function through a decorator, this internal information is lost, especially in cases where type variables are involved. That's because the application of a generic decorator must transform the generic return type into a concrete return type as part of "solving" the type variable.

Here's a workaround that uses a `ParamSpec`. This has the added benefit of working with any property in any class rather than being constrained to only the `Parent` class and its subclasses.

```python
P = ParamSpec("P")

def handle_error(fn: Callable[P, float]) -> Callable[P, float]:
    def wrapper(*args: P.args, **kwargs: P.kwargs) -> float:
        try:
            return fn(*args, **kwargs)
        except ZeroDivisionError:
            return 0

    return wrapper
```

**Reply by argilya** [reply]
Thank you!
